### PR TITLE
jerry: implement source info

### PR DESCRIPTION
**Pull request by algebrait**
This PR implements source file name, function line and column number in vm. Snapshot is also supported.
Line and column info only support start position of function as without this PR. Detailed info would be implemented later.

##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
    benchmark/error/stack.js

|  | line-info | base |
| ------ | ------ | ------ |
| 1 | 2,067 | 1,615 |
| 2 | 2,102 | 1,626 |
| 3 | 2,092 | 1,631 |
| avg | 2,087 | 1,624 |
| + | 28% |  |

if download [es6-shim](https://raw.githubusercontent.com/paulmillr/es6-shim/master/es6-shim.js) to benchmark/error/ and add

```
require('es6-shim')
```
in benchmark/error/stack.js, result is following

|  | line-info | base |
| ------ | ------ | ------ |
| 1 | 1,612 | 84 |
| 2 | 1,600 | 84 |
| 3 | 1,607 | 84 |
| avg | 1606 | 84 |
| + | 1812% |  |

- [x] documentation is changed or added
    Jerry api changed which caused following docs updated,
    02.API-REFERENCE.md
    03.API-EXAMPLE.md
- [x] CPU profiler

### CODE REVIEW

**Comment by algebrait** [comment]
Not completed entirely. Just for review.

File: deps/jerry/jerry-core/CMakeLists.txt [hunk 1]
```diff
@@ -26,7 +26,7 @@ set(FEATURE_MEM_STATS          OFF      CACHE BOOL   "Enable memory statistics?"
 set(FEATURE_MEM_STRESS_TEST    OFF      CACHE BOOL   "Enable mem-stress test?")
 set(FEATURE_CPU_PROFILER       OFF      CACHE BOOL   "Enable Cpu profiler?")
 set(FEATURE_HEAP_PROFILER      OFF      CACHE BOOL   "Enable Heap profiler?")
-set(FEATURE_FUNCTION_NAME      OFF      CACHE BOOL   "Enable saving Javascript function name?")
+set(FEATURE_DEBUG_INFO         ON       CACHE BOOL   "Enable saving Javascript debug info?")
```
Comment (yorkie) [inline deps/jerry/jerry-core/CMakeLists.txt]: Suggestion name: FEATURE_EXTRA_INFO, the _DEBUG_ might be confusion with the jerry-debugger and the define `JERRY_NDEBUG`.

Comment (algebrait) [reply deps/jerry/jerry-core/CMakeLists.txt]: EXTRA_INFO is too generic. DEBUG_SYMBOL is ok?

Comment (yorkie) [reply deps/jerry/jerry-core/CMakeLists.txt]: Good to me :)

File: deps/jerry/jerry-core/ecma/operations/ecma-exceptions.c [hunk 2]
```diff
@@ -144,37 +144,13 @@ ecma_new_standard_error (ecma_standard_error_t error_type) /**< native error typ
                                                                          frames_str_p,
                                                                          ECMA_PROPERTY_CONFIGURABLE_WRITABLE,
                                                                          NULL);
+  ecma_value_t frame = jerry_get_backtrace ();
+  ecma_named_data_property_assign_value (new_error_obj_p,
+		                                 prop_value_p,
+		                                 frame);
```
Comment (yorkie) [inline deps/jerry/jerry-core/ecma/operations/ecma-exceptions.c]: Alignments is required.

File: deps/jerry/jerry-core/parser/js/js-parser-internal.h [hunk 3]
```diff
@@ -272,6 +272,7 @@ typedef struct
   /* Lexer members. */
   lexer_token_t token;                        /**< current token */
   lexer_lit_object_t lit_object;              /**< current literal object */
+  ecma_value_t resource_name;
```
Comment (yorkie) [inline deps/jerry/jerry-core/parser/js/js-parser-internal.h]: Comment to describe this field is required :)

File: deps/jerry/jerry-core/api/jerry-snapshot.c [hunk 4]
```diff
@@ -697,7 +705,9 @@ typedef enum
  *         0 - otherwise.
  */
 static size_t
-jerry_parse_and_save_snapshot_with_args (const jerry_char_t *source_p, /**< script source */
+jerry_parse_and_save_snapshot_with_args (const jerry_char_t *resource_name_p, /**< resource name (usually a file name) */
```
[the remark this answers is not in the document]
Comment (algebrait) [reply deps/jerry/jerry-core/api/jerry-snapshot.c]: "resource name" follows jerry. Although usually a file name, any name such as module id is accepted.

[the remark this answers is not in the document]
Comment (yorkie) [reply deps/jerry/jerry-core/api/jerry-snapshot.c]: Looks good to me :)

File: deps/jerry/jerry-core/CMakeLists.txt (hunk 1 above)
Comment (algebrait) [reply deps/jerry/jerry-core/CMakeLists.txt]: If DEBUG_SYMBOL is still confusing with NDEBUG, is SOURCE_SYMBOL ok?

**Comment by algebrait** [comment]
Performance improved 28% on benchmark/error/stack.js.
This benchmark is too simple. Complicated case should be introduced to show great performance improvement.

File: deps/jerry/tools/profiler/j2cl.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+
+/* Convert jerry CPU profiler data into collapse stack file.
+ * It convert every input line
+ * time,function1(source1:line1:column1),...,functionn(sourcen:linen:columnn)
+ * to line
+ * functionn(sourcen:linen:columnn);...;function1(source1:line1:column1) time
+ * Call graph is functionn->...->function1
```
Comment (legendecas) [inline deps/jerry/tools/profiler/j2cl.js]: looks like comment has not been terminated.

Comment (algebrait) [reply deps/jerry/tools/profiler/j2cl.js]: 3x. */ added

**Pull request merged**